app-header/app-header.html

```diff
@@ -396,42 +413,77 @@
         if (this._height === 0) {
           return;
         }
-        var progress;
-        var mayHeaderMove = !this.fixed || this.condenses;
-        var top = Math.min(this._headerDy, Math.max(0,
-            this.reveals ? this._top + scrollTop - this._lastScrollTop : scrollTop
-        ));
-
-        if (this._shouldCondense(scrollTop)) {
-          top = Math.max(this._deltaHeight, top);
+
+        var progress = 0;
+        var top = 0;
+        var lastTop = this._top;
+        var lastScrollTop = this._lastScrollTop;
+        var maxHeaderTop = this._maxHeaderTop;
+        var dScrollTop = scrollTop - this._lastScrollTop;
+        var absDScrollTop = Math.abs(dScrollTop);
+        var isScrollingDown = scrollTop > this._lastScrollTop;
+        var now = Date.now();
+
+        if (this._mayMove()) {
+          top = this._clamp(this.reveals ? lastTop + dScrollTop : scrollTop, 0, maxHeaderTop);
+        }
+
+        if (scrollTop >= this._dHeight) {
+          top = this.condenses ? Math.max(this._dHeight, top) : top;
+          this.style.transitionDuration = '0ms';
+        }
+
+        if (this.reveals && !this.disabled && absDScrollTop < 100) {
+          // set the initial scroll position
+          if (now - this._initTimestamp > 300 || this._wasScrollingDown !== isScrollingDown) {
+            this._initScrollTop = scrollTop;
+            this._initTimestamp = now;
+          }
+
+          if (scrollTop >= maxHeaderTop) {
+            // check if the header is allowed to snap
+            if (Math.abs(this._initScrollTop - scrollTop) > 30 || absDScrollTop > 10) {
+              if (isScrollingDown && scrollTop >= maxHeaderTop) {
+                top = maxHeaderTop;
+              } else if (!isScrollingDown && scrollTop >= this._dHeight) {
+                top = this.condenses ? this._dHeight : 0;
+              }
+
+              var scrollVelocity = dScrollTop / (now - this._lastTimestamp);
+              this.style.transitionDuration = this._clamp((top - lastTop) / scrollVelocity , 0, 300) + 'ms';
+            } else {
+              top = this._top;
+            }
+          }
         }
-        if (this._deltaHeight === 0) {
+
+        if (this._dHeight === 0) {
           progress = scrollTop > 0 ? 1 : 0;
         } else {
-          progress = top / this._deltaHeight;
-        }
-        if (forceUpdate || progress !== this._progress || top !== this._top || scrollTop === 0) {
-          this._runEffects(progress, mayHeaderMove ? top : 0);
-          this._transformHeader(mayHeaderMove ? top : 0);
-          this.fire('app-header-transform', { top: top, progress: progress }, { bubbles: false });
```

Comment:
There is still one more demo using `app-header-transform`: app-header-layout/demo/music.html